thirdparty/nginx/nginx-1.7.1-zimbra/src/mail/ngx_mail_proxy_module.c

```diff
@@ -1742,6 +1742,13 @@ ngx_mail_proxy_upstream_error(ngx_mail_session_t *s)
         ngx_log_debug1(NGX_LOG_DEBUG_MAIL, s->connection->log, 0,
                        "close mail proxy connection: %d",
                        s->proxy->upstream.connection->fd);
+#if (NGX_MAIL_SSL)
+
+        if (s->proxy->upstream.connection->ssl) {
+            s->proxy->upstream.connection->ssl->no_wait_shutdown = 1;
+            ngx_ssl_shutdown(s->proxy->upstream.connection);
```

Comment:
Appropriate action should be taken based on the return value of ngx_ssl_shutdown.

Reply:
@thefoolwhocodes The patch consists merely of applying the conditional code already present at 
https://github.com/Zimbra/packages/blob/42c19f9bd20e097f63d2e6f050ac8d35a44a05f9/thirdparty/nginx/nginx-1.7.1-zimbra/src/mail/ngx_mail_proxy_module.c#L1774-L1797
to the other functions.
Could you provide an example of what that appropriate action would be, since I guess it's missing in the code path functioning properly too?

Reply:
If we are so sure as we already are, that we are always expecting NGX_OK and there is no way the function is gonna return NGX_ERROR or NGX_AGAIN, please use the negation of same, if the function does not return what we are expecting out of it and log the same.
Like
if (NGX_OK != function call)
{
	//Log in warn that it is not behaving as expected
}

A small comment above the if block should suffice why we think that we don't need to handle NGX_ERROR or NGX_AGAIN conditions from the call.

Reply:
One more suggestion:
- Please use the capabilities of no_send_shutdown too if you do not want to notify the peer about the shutdown, as our goal is just to free the structure.

Reply:
Hello @thefoolwhocodes 
I was looking at other functions in the nginx codebase to see how the same thing was accomplished.
Would something along the lines of this
https://github.com/Zimbra/packages/blob/42c19f9bd20e097f63d2e6f050ac8d35a44a05f9/thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_keepalive_module.c#L410-L430
be appropriate here as well, in your opinion? (I am referring to ssl-related part).

Reply:
@zx-fulvio looks fine, few things to consider
- If you cannot fit  NGX_AGAIN from the above example in your code, you could try doing this:
connection->ssl->no_wait_shutdown = 1;
connection->ssl->no_send_shutdown = 1;
(void) ngx_ssl_shutdown(connection);
- Also try nullifying the connection after connection close in code flow.

Reply:
So something like
```c
static void
ngx_mail_proxy_upstream_error(ngx_mail_session_t *s)
{
    if (s->proxy->upstream.connection) {
        ngx_log_debug1(NGX_LOG_DEBUG_MAIL, s->connection->log, 0,
                       "close mail proxy connection: %d",
                       s->proxy->upstream.connection->fd);
#if (NGX_MAIL_SSL)

        if (s->proxy->upstream.connection->ssl) {
            s->proxy->upstream.connection->ssl->no_wait_shutdown = 1;
            s->proxy->upstream.connection->ssl->no_send_shutdown = 1;

            if (ngx_ssl_shutdown(s->proxy->upstream.connection) == NGX_AGAIN) {
                s->proxy->upstream.connection->ssl->handler = (ngx_connection_handler_pt) ngx_mail_proxy_upstream_error;
                return;
            }
        }

#endif

        ngx_close_connection(s->proxy->upstream.connection);
        s->proxy->upstream.connection = NULL;
    }

    if (s->out.len == 0) {
        ngx_mail_session_internal_server_error(s);
        return;
    }

    s->quit = 1;
    ngx_mail_send(s->connection->write);
}
```
the code above?

Reply:
Looks fine..but you got to do some testing over the handler's behavior once you get a NGX_AGAIN from  ngx_ssl_shutdown.